File: src/main/resources/kubejs_export/gtponder_api.js
"use strict";

/*
 * GTPonder simplified PonderJS helper API.
 *
 * This file is for pack and addon authors who want to write approachable
 * hand-authored Ponder scenes without copying the larger generated multiblock
 * scene machinery. It deliberately wraps only common operations: positions,
 * block placement, text, outlines, item cues, flow lines, and simple reveals.
 *
 * Example:
 *
 * Ponder.registry((event) => {
 *   GTPonder.scene(event, "gtceu:lv_mixer", "my_mod:mixer/example", "Mixer Basics", (p) => {
 *     let mixer = p.pos(3, 1, 3);
 *     p.block(mixer, "gtceu:lv_mixer");
 *     p.text("The mixer combines item and fluid ingredients.", mixer, {
 *       palette: "input",
 *     });
 *     p.idle(50);
 *   });
 * });
 */

const $GTPonderBlockPos = Java.loadClass("net.minecraft.core.BlockPos");
const $GTPonderDirection = Java.loadClass("net.minecraft.core.Direction");
const $GTPonderBlocks = Java.loadClass("net.minecraft.world.level.block.Blocks");
const $GTPonderItemStack = Java.loadClass("net.minecraft.world.item.ItemStack");
const $GTPonderPonderPalette = Java.loadClass(
  "net.createmod.ponder.api.PonderPalette",
);
const $GTPonderPointing = Java.loadClass("net.createmod.catnip.math.Pointing");
const $GTPonderForgeRegistries = Java.loadClass(
  "net.minecraftforge.registries.ForgeRegistries",
);
const $GTPonderResourceLocation = Java.loadClass(
  "net.minecraft.resources.ResourceLocation",
);

const GTPONDER_API_VERSION = "0.1.0";
const GTPONDER_DEFAULT_STRUCTURE_ID = new $GTPonderResourceLocation(
  "tfg:gregtech_multiblocks/blank_64",
);
const GTPONDER_DEFAULT_BASE_PLATE_SIZE = 7;
const GTPONDER_DEFAULT_TEXT_DURATION = 56;
const GTPONDER_DEFAULT_CUE_DURATION = 36;
const GTPONDER_DEFAULT_FLOW_DURATION = 34;
const GTPONDER_DEFAULT_REVEAL_IDLE = 2;

const GTPONDER_PALETTES = {
  white: $GTPonderPonderPalette.WHITE,
  black: $GTPonderPonderPalette.BLACK,
  red: $GTPonderPonderPalette.RED,
  green: $GTPonderPonderPalette.GREEN,
  blue: $GTPonderPonderPalette.BLUE,
  slow: $GTPonderPonderPalette.SLOW,
  medium: $GTPonderPonderPalette.MEDIUM,
  fast: $GTPonderPonderPalette.FAST,
  input: $GTPonderPonderPalette.INPUT,
  output: $GTPonderPonderPalette.OUTPUT,
};

const GTPONDER_DIRECTIONS = {
  up: $GTPonderDirection.UP,
  down: $GTPonderDirection.DOWN,
  north: $GTPonderDirection.NORTH,
  south: $GTPonderDirection.SOUTH,
  west: $GTPonderDirection.WEST,
  east: $GTPonderDirection.EAST,
};

const GTPONDER_POINTING = {
  up: $GTPonderPointing.UP,
  down: $GTPonderPointing.DOWN,
  left: $GTPonderPointing.LEFT,
  right: $GTPonderPointing.RIGHT,
};

function gtPonderOptions(options) {
  return options === null || options === undefined ? {} : options;
}

function gtPonderNumber(value, fallback) {
  let numberValue = Number(value);
  return isNaN(numberValue) ? fallback : numberValue;
}

function gtPonderResourceLocation(id) {
  if (id instanceof $GTPonderResourceLocation) {
    return id;
  }

  return new $GTPonderResourceLocation(String(id));
}

function gtPonderPos(x, y, z) {
  if (Array.isArray(x)) {
    return new $GTPonderBlockPos(x[0], x[1], x[2]);
  }

  if (x !== null && x !== undefined && typeof x.getX === "function") {
    return x;
  }

  return new $GTPonderBlockPos(x, y, z);
}

function gtPonderBlockState(id) {
  let block = $GTPonderForgeRegistries.BLOCKS.getValue(
    gtPonderResourceLocation(id),
  );
  if (block === null) {
    console.warn(`Unknown block in GTPonder scene: ${id}`);
    return $GTPonderBlocks.BARRIER.defaultBlockState();
  }

  return block.defaultBlockState();
}

function gtPonderItemStack(id) {
  if (id instanceof $GTPonderItemStack) {
    return id;
  }

  let item = $GTPonderForgeRegistries.ITEMS.getValue(
    gtPonderResourceLocation(id),
  );
  if (item === null) {
    console.warn(`Unknown item in GTPonder scene: ${id}`);
    item = $GTPonderForgeRegistries.ITEMS.getValue(
      gtPonderResourceLocation("minecraft:barrier"),
    );
  }

  return new $GTPonderItemStack(item);
}

function gtPonderPalette(value, fallback) {
  if (value === null || value === undefined) {
    return fallback || $GTPonderPonderPalette.WHITE;
  }

  let key = String(value).toLowerCase();
  return GTPONDER_PALETTES[key] || value;
}

function gtPonderDirection(value, fallback) {
  if (value === null || value === undefined) {
    return fallback || $GTPonderDirection.DOWN;
  }

  let key = String(value).toLowerCase();
  return GTPONDER_DIRECTIONS[key] || value;
}

function gtPonderPointing(value, fallback) {
  if (value === null || value === undefined) {
    return fallback || $GTPonderPointing.DOWN;
  }

  let key = String(value).toLowerCase();
  return GTPONDER_POINTING[key] || value;
}

function gtPonderBlockCenter(util, pos) {
  let blockPos = gtPonderPos(pos);
  return util.vector.of(
    blockPos.getX() + 0.5,
    blockPos.getY() + 0.5,
    blockPos.getZ() + 0.5,
  );
}

function gtPonderBlockTop(util, pos, xOffset, zOffset) {
  let blockPos = gtPonderPos(pos);
  return util.vector.of(
    blockPos.getX() + 0.5 + gtPonderNumber(xOffset, 0),
    blockPos.getY() + 1.05,
    blockPos.getZ() + 0.5 + gtPonderNumber(zOffset, 0),
  );
}

function gtPonderSelectionForPositions(util, positions) {
  let selection = null;

  positions.forEach((entry) => {
    let next = util.select.position(gtPonderPos(entry));
    selection = selection === null ? next : selection.add(next);
  });

  return selection;
}

function gtPonderNormalizeBlockEntry(entry) {
  if (Array.isArray(entry)) {
    return { pos: gtPonderPos(entry[0], entry[1], entry[2]), id: entry[3] };
  }

  return {
    pos: gtPonderPos(entry.pos || entry.position),
    id: entry.id || entry.block,
  };
}

function gtPonderContext(scene, util, sceneOptions) {
  let options = gtPonderOptions(sceneOptions);
  let defaultTextDuration =
    options.textDuration || GTPONDER_DEFAULT_TEXT_DURATION;
  let defaultCueDuration = options.cueDuration || GTPONDER_DEFAULT_CUE_DURATION;
  let defaultFlowDuration =
    options.flowDuration || GTPONDER_DEFAULT_FLOW_DURATION;

  function showBlock(pos, id, blockOptions) {
    let localOptions = gtPonderOptions(blockOptions);
    let blockPos = gtPonderPos(pos);
    scene.world.setBlock(blockPos, gtPonderBlockState(id), false);

    if (localOptions.show !== false) {
      scene.world.showSection(
        util.select.position(blockPos),
        gtPonderDirection(localOptions.direction, $GTPonderDirection.DOWN),
      );
    }

    return blockPos;
  }

  function showBlocks(blocks, blockOptions) {
    let localOptions = gtPonderOptions(blockOptions);
    let positions = [];

    blocks.forEach((entry) => {
      let block = gtPonderNormalizeBlockEntry(entry);
      scene.world.setBlock(block.pos, gtPonderBlockState(block.id), false);
      positions.push(block.pos);
    });

    let selection = gtPonderSelectionForPositions(util, positions);
    if (selection !== null && localOptions.show !== false) {
      scene.world.showSection(
        selection,
        gtPonderDirection(localOptions.direction, $GTPonderDirection.DOWN),
      );
    }

    return positions;
  }

  function showText(text, targetPos, textOptions) {
    let localOptions = gtPonderOptions(textOptions);
    let duration = localOptions.duration || defaultTextDuration;
    let overlay = scene.overlay
      .showText(duration)
      .text(text)
      .colored(
        gtPonderPalette(localOptions.palette || localOptions.color, GTPONDER_PALETTES.white),
      );

    if (targetPos !== null && targetPos !== undefined) {
      overlay.pointAt(gtPonderBlockCenter(util, targetPos));
      if (localOptions.placeNearTarget !== false) {
        overlay.placeNearTarget();
      }
    }

    return overlay;
  }

  function showOutline(key, targetPos, outlineOptions) {
    let localOptions = gtPonderOptions(outlineOptions);
    let duration = localOptions.duration || defaultTextDuration;
    let palette = gtPonderPalette(
      localOptions.palette || localOptions.color,
      GTPONDER_PALETTES.blue,
    );

    scene.overlay.showOutline(
      palette,
      key,
      util.select.position(gtPonderPos(targetPos)),
      duration,
    );
  }

  function showLine(fromPos, toPos, lineOptions) {
    let localOptions = gtPonderOptions(lineOptions);
    let duration = localOptions.duration || defaultFlowDuration;
    let palette = gtPonderPalette(
      localOptions.palette || localOptions.color,
      GTPONDER_PALETTES.medium,
    );

    if (localOptions.big === false) {
      scene.overlay.showLine(
        palette,
        gtPonderBlockCenter(util, fromPos),
        gtPonderBlockCenter(util, toPos),
        duration,
      );
      return;
    }

    scene.overlay.showBigLine(
      palette,
      gtPonderBlockCenter(util, fromPos),
      gtPonderBlockCenter(util, toPos),
      duration,
    );
  }

  function showItemCue(itemId, targetPos, cueOptions) {
    let localOptions = gtPonderOptions(cueOptions);
    let cue = scene.overlay.showControls(
      gtPonderBlockTop(util, targetPos, localOptions.xOffset, localOptions.zOffset),
      gtPonderPointing(localOptions.pointing, $GTPonderPointing.DOWN),
      localOptions.duration || defaultCueDuration,
    );

    if (localOptions.rightClick === true) {
      cue.rightClick();
    }

    cue.withItem(gtPonderItemStack(itemId));
  }

  function showItemEntity(itemId, targetPos, entityOptions) {
    let localOptions = gtPonderOptions(entityOptions);
    scene.world.createItemEntity(
      gtPonderBlockTop(util, targetPos, localOptions.xOffset, localOptions.zOffset),
      util.vector.of(
        gtPonderNumber(localOptions.velocityX, 0),
        gtPonderNumber(localOptions.velocityY, 0.02),
        gtPonderNumber(localOptions.velocityZ, 0),
      ),
      gtPonderItemStack(itemId),
    );
  }

  function reveal(blocks, revealOptions) {
    let localOptions = gtPonderOptions(revealOptions);
    let idleTime = localOptions.idle || GTPONDER_DEFAULT_REVEAL_IDLE;
    let direction = gtPonderDirection(localOptions.direction, $GTPonderDirection.DOWN);

    blocks.forEach((entry) => {
      showBlocks([entry], { direction: direction });
      scene.idle(idleTime);
    });
  }

  return {
    scene: scene,
    util: util,
    palette: GTPONDER_PALETTES,
    direction: GTPONDER_DIRECTIONS,
    pointing: GTPONDER_POINTING,
    pos: gtPonderPos,
    center: (pos) => gtPonderBlockCenter(util, pos),
    top: (pos, xOffset, zOffset) =>
      gtPonderBlockTop(util, pos, xOffset, zOffset),
    blockState: gtPonderBlockState,
    itemStack: gtPonderItemStack,
    selection: (positions) => gtPonderSelectionForPositions(util, positions),
    idle: (ticks) => scene.idle(ticks),
    block: showBlock,
    blocks: showBlocks,
    text: showText,
    outline: showOutline,
    highlight: showOutline,
    line: showLine,
    flow: showLine,
    itemCue: showItemCue,
    rightClick: (itemId, targetPos, cueOptions) => {
      let localOptions = gtPonderOptions(cueOptions);
      localOptions.rightClick = true;
      showItemCue(itemId, targetPos, localOptions);
    },
    itemEntity: showItemEntity,
    reveal: reveal,
  };
}

function gtPonderScene(event, target, sceneId, title, options, callback) {
  let sceneOptions = options;
  let sceneCallback = callback;

  if (typeof options === "function") {
    sceneOptions = {};
    sceneCallback = options;
  }

  sceneOptions = gtPonderOptions(sceneOptions);

  event.create(target).scene(
    sceneId,
    title,
    sceneOptions.structure || GTPONDER_DEFAULT_STRUCTURE_ID,
    (scene, util) => {
      let basePlateSize =
        sceneOptions.basePlateSize ||
        sceneOptions.basePlate ||
        GTPONDER_DEFAULT_BASE_PLATE_SIZE;
      scene.configureBasePlate(
        sceneOptions.basePlateX || 0,
        sceneOptions.basePlateZ || 0,
        basePlateSize,
      );

      if (sceneOptions.scale !== null && sceneOptions.scale !== undefined) {
        scene.scaleSceneView(sceneOptions.scale);
      }
      if (sceneOptions.offsetY !== null && sceneOptions.offsetY !== undefined) {
        scene.setSceneOffsetY(sceneOptions.offsetY);
      }
      if (sceneOptions.showBasePlate !== false) {
        scene.showBasePlate();
      }
      if (sceneOptions.initialIdle !== false) {
        scene.idle(sceneOptions.initialIdle || 3);
      }

      sceneCallback(gtPonderContext(scene, util, sceneOptions));

      if (sceneOptions.markAsFinished !== false) {
        scene.markAsFinished();
      }
    },
  );
}

global.GTPonder = {
  version: GTPONDER_API_VERSION,
  defaultStructureId: GTPONDER_DEFAULT_STRUCTURE_ID,
  palette: GTPONDER_PALETTES,
  direction: GTPONDER_DIRECTIONS,
  pointing: GTPONDER_POINTING,
  pos: gtPonderPos,
  resourceLocation: gtPonderResourceLocation,
  blockState: gtPonderBlockState,
  itemStack: gtPonderItemStack,
  scene: gtPonderScene,
  context: gtPonderContext,
};

console.info(`Loaded GTPonder helper API ${GTPONDER_API_VERSION}.`);

File: src/main/java/net/phoenix/core/integration/ponder/api/GTPonderAPI.java
package net.phoenix.core.integration.ponder.api;

import net.createmod.catnip.math.Pointing;
import net.createmod.ponder.api.PonderPalette;
import net.createmod.ponder.api.scene.SceneBuildingUtil;
import net.createmod.ponder.api.scene.Selection;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.phys.Vec3;
import net.minecraftforge.registries.ForgeRegistries;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class GTPonderAPI {

    private static final Logger LOGGER = LogManager.getLogger("GTPonderAPI");

    public static final String GTPONDER_API_VERSION = "0.1.0";
    public static final ResourceLocation GTPONDER_DEFAULT_STRUCTURE_ID = new ResourceLocation(
            "phoenixcore:blank_48");
    public static final int GTPONDER_DEFAULT_BASE_PLATE_SIZE = 7;
    public static final int GTPONDER_DEFAULT_TEXT_DURATION = 56;
    public static final int GTPONDER_DEFAULT_CUE_DURATION = 36;
    public static final int GTPONDER_DEFAULT_FLOW_DURATION = 34;
    public static final int GTPONDER_DEFAULT_REVEAL_IDLE = 2;

    public static final Map<String, PonderPalette> GTPONDER_PALETTES = new HashMap<>();
    public static final Map<String, Direction> GTPONDER_DIRECTIONS = new HashMap<>();
    public static final Map<String, Pointing> GTPONDER_POINTING = new HashMap<>();

    static {
        // Initialize GTPONDER_PALETTES
        GTPONDER_PALETTES.put("white", PonderPalette.WHITE);
        GTPONDER_PALETTES.put("black", PonderPalette.BLACK);
        GTPONDER_PALETTES.put("red", PonderPalette.RED);
        GTPONDER_PALETTES.put("green", PonderPalette.GREEN);
        GTPONDER_PALETTES.put("blue", PonderPalette.BLUE);
        GTPONDER_PALETTES.put("slow", PonderPalette.SLOW);
        GTPONDER_PALETTES.put("medium", PonderPalette.MEDIUM);
        GTPONDER_PALETTES.put("fast", PonderPalette.FAST);
        GTPONDER_PALETTES.put("input", PonderPalette.INPUT);
        GTPONDER_PALETTES.put("output", PonderPalette.OUTPUT);

        // Initialize GTPONDER_DIRECTIONS
        GTPONDER_DIRECTIONS.put("up", Direction.UP);
        GTPONDER_DIRECTIONS.put("down", Direction.DOWN);
        GTPONDER_DIRECTIONS.put("north", Direction.NORTH);
        GTPONDER_DIRECTIONS.put("south", Direction.SOUTH);
        GTPONDER_DIRECTIONS.put("west", Direction.WEST);
        GTPONDER_DIRECTIONS.put("east", Direction.EAST);

        // Initialize GTPONDER_POINTING
        GTPONDER_POINTING.put("up", Pointing.UP);
        GTPONDER_POINTING.put("down", Pointing.DOWN);
        GTPONDER_POINTING.put("left", Pointing.LEFT);
        GTPONDER_POINTING.put("right", Pointing.RIGHT);
    }

    public static Map<String, Object> gtPonderOptions(Map<String, Object> options) {
        return options == null ? new HashMap<>() : options;
    }

    public static double gtPonderNumber(Object value, double fallback) {
        if (value instanceof Number) {
            return ((Number) value).doubleValue();
        }
        try {
            double numberValue = Double.parseDouble(String.valueOf(value));
            return Double.isNaN(numberValue) ? fallback : numberValue;
        } catch (NumberFormatException e) {
            return fallback;
        }
    }

    public static ResourceLocation gtPonderResourceLocation(Object id) {
        if (id instanceof ResourceLocation) {
            return (ResourceLocation) id;
        }
        return new ResourceLocation(String.valueOf(id));
    }

    public static BlockPos gtPonderPos(Object x, Object y, Object z) {
        if (x instanceof List) {
            List<?> list = (List<?>) x;
            if (list.size() >= 3) {
                return new BlockPos(
                        (int) gtPonderNumber(list.get(0), 0),
                        (int) gtPonderNumber(list.get(1), 0),
                        (int) gtPonderNumber(list.get(2), 0));
            }
        }
        if (x instanceof BlockPos) {
            return (BlockPos) x;
        }
        if (x != null && x.getClass().getName().contains("BlockPos")) {
            try {
                int bx = (int) x.getClass().getMethod("getX").invoke(x);
                int by = (int) x.getClass().getMethod("getY").invoke(x);
                int bz = (int) x.getClass().getMethod("getZ").invoke(x);
                return new BlockPos(bx, by, bz);
            } catch (Exception e) {
                LOGGER.warn("Failed to convert KJS BlockPos to Java BlockPos: " + e.getMessage());
            }
        }
        return new BlockPos(
                (int) gtPonderNumber(x, 0),
                (int) gtPonderNumber(y, 0),
                (int) gtPonderNumber(z, 0));
    }

    public static BlockState gtPonderBlockState(Object id) {
        ResourceLocation rl = gtPonderResourceLocation(id);
        if (ForgeRegistries.BLOCKS.containsKey(rl)) {
            return ForgeRegistries.BLOCKS.getValue(rl).defaultBlockState();
        }
        LOGGER.warn("Unknown block in GTPonder scene: " + id);
        return Blocks.BARRIER.defaultBlockState();
    }

    public static ItemStack gtPonderItemStack(Object id) {
        if (id instanceof ItemStack) {
            return (ItemStack) id;
        }
        ResourceLocation rl = gtPonderResourceLocation(id);
        if (ForgeRegistries.ITEMS.containsKey(rl)) {
            return new ItemStack(ForgeRegistries.ITEMS.getValue(rl));
        }
        LOGGER.warn("Unknown item in GTPonder scene: " + id);
        return new ItemStack(Blocks.BARRIER.asItem());
    }

    public static PonderPalette gtPonderPalette(Object value, PonderPalette fallback) {
        if (value == null) {
            return fallback != null ? fallback : PonderPalette.WHITE;
        }
        String key = String.valueOf(value).toLowerCase();
        return GTPONDER_PALETTES.getOrDefault(key, fallback != null ? fallback : PonderPalette.WHITE);
    }

    public static Direction gtPonderDirection(Object value, Direction fallback) {
        if (value == null) {
            return fallback != null ? fallback : Direction.DOWN;
        }
        String key = String.valueOf(value).toLowerCase();
        return GTPONDER_DIRECTIONS.getOrDefault(key, fallback != null ? fallback : Direction.DOWN);
    }

    public static Pointing gtPonderPointing(Object value, Pointing fallback) {
        if (value == null) {
            return fallback != null ? fallback : Pointing.DOWN;
        }
        String key = String.valueOf(value).toLowerCase();
        return GTPONDER_POINTING.getOrDefault(key, fallback != null ? fallback : Pointing.DOWN);
    }

    public static Vec3 gtPonderBlockCenter(SceneBuildingUtil util, BlockPos pos) {
        return util.vector().of(pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5);
    }

    public static Vec3 gtPonderBlockTop(SceneBuildingUtil util, BlockPos pos, Object xOffset, Object zOffset) {
        return util.vector().of(
                pos.getX() + 0.5 + gtPonderNumber(xOffset, 0),
                pos.getY() + 1.05,
                pos.getZ() + 0.5 + gtPonderNumber(zOffset, 0));
    }

    public static Selection gtPonderSelectionForPositions(SceneBuildingUtil util, List<BlockPos> positions) {
        Selection selection = null;
        for (BlockPos pos : positions) {
            Selection next = util.select().position(pos);
            selection = (selection == null) ? next : selection.add(next);
        }
        return selection;
    }

    public static BlockEntry gtPonderNormalizeBlockEntry(Object entry) {
        if (entry instanceof List) {
            List<?> list = (List<?>) entry;
            if (list.size() >= 4) {
                return new BlockEntry(gtPonderPos(list.get(0), list.get(1), list.get(2)), list.get(3));
            }
        } else if (entry instanceof Map) {
            Map<?, ?> map = (Map<?, ?>) entry;
            Object posObj = map.get("pos");
            if (posObj == null) posObj = map.get("position");
            Object idObj = map.get("id");
            if (idObj == null) idObj = map.get("block");
            return new BlockEntry(gtPonderPos(posObj, null, null), idObj);
        }
        throw new IllegalArgumentException("Invalid block entry format: " + entry);
    }

    public static class BlockEntry {

        public final BlockPos pos;
        public final Object id; // Can be ResourceLocation or String

        public BlockEntry(BlockPos pos, Object id) {
            this.pos = pos;
            this.id = id;
        }
    }
}
